pkg/rpcmiddleware/rpcretry/stream_client.go

```diff
@@ -0,0 +1,148 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcretry
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type retriableStreamClient struct {
+	grpc.ClientStream
+	mu            sync.Mutex
+	bufferedSends []interface{}
+	streamerCall  func(ctx context.Context) (grpc.ClientStream, error)
+	counter       uint
+
+	callOpts  *options
+	parentCtx context.Context
+}
+
+func (s *retriableStreamClient) setStream(clientStream grpc.ClientStream) {
+	s.mu.Lock()
+	s.ClientStream = clientStream
+	s.mu.Unlock()
+}
+
+func (s *retriableStreamClient) getStream() grpc.ClientStream {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.ClientStream
+}
+
+func (s *retriableStreamClient) shouldRetry(err error) bool {
+	if err == nil || err == io.EOF {
+		return false
+	}
+	return isRetriable(err, s.callOpts)
+}
+
+func (s *retriableStreamClient) incrementCounter() uint {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.counter++
+	return s.counter
+}
+
+func (s *retriableStreamClient) SendMsg(m interface{}) error {
+	s.mu.Lock()
+	s.bufferedSends = append(s.bufferedSends, m)
```

Comment:
This looks like a memory leak waiting to happen.

Reply:
True. Will make it so that when attempting to reconnect it will only send the last send message, that way it shouldn't really be a problem.

Reply:
I didn't notice the obvious before pushing the change but the biggest problem of not buffering the all the previous messages would be that a possible send that causes the error can easily be lost. 
For example:
send msgA (causes error) -> send msgB -> received error from msgA (uses msgB content to retry).

What I'm wondering is how to mitigate that in a sensible manner, I'm thinking of just using channels to store the message until the `RecvMsg` returns a nonRetriable error (that includes the closure of the stream) and then discard the message

Reply:
@htdvisser I'll admit that I was thrown in for a loop, especially since I'm thinking in terms of a bidiStream. I might be mistaken but after reading the code and revising my knowledge on the gRPC communications, the memory leak would only happen in a BidiStream where the client can send more than one message. In a `1:n` communication that problem wouldn't happen.

With that in mind, it makes sense to keep the new change where is no longer a slice but just an `interface{}`, it would be a problem in the future when trying to do an interceptor that works for bidirectional stream but I don't think that is going to be the case soon.

Reply:
I still think we should only do retries for unary RPCs, and remove the streaming interceptor completely.

It's not difficult to come up with an example for this going wrong in a server-streaming RPC (`1:n`). Take for example our `events.Stream` RPC. The client sends the stream request, including `after=2021-01-26T00:00:00Z`. At some point the stream breaks, the client retries, and suddenly receives duplicate messages that it already received before the stream broke. This is unexpected behavior, and instead I think the client should implement its own retry logic; in this case, retry the request with `after` set to the timestamp of the last event it did receive.

I imagine there will be a lot of other examples like this, so I just don't think that it's possible to build a one-size-fits-all interceptor for streaming RPCs.